### DISCUSSION TITLE
[Templates] Best way to output a variable containing a list of strings as a TOML array of strings

### DISCUSSION THREAD

**Discussion by bradenhilton** [twpayne/chezmoi#2119] [Q&A] [answered]
Currently, I have this in my config template:

```
{{ $pythonCmd := "" -}}
{{ if lookPath "python3" -}}
{{   $pythonCmd = "python3" -}}
{{ else if lookPath "python" -}}
{{   $pythonCmd = "python" -}}
{{ end -}}
{{ if ne "" $pythonCmd -}}
[interpreters.py]
command = {{ $pythonCmd | quote }}

{{ end -}}
```

I'd like to use `py -3 ...` if it's available in my path, so I altered the template to:

```
{{ $pythonCmd := "" -}}
{{ $pythonArgs := list "" -}}
{{ if lookPath "py" -}}
{{   $pythonCmd = "py" -}}
{{   $pythonArgs = list "-3" "test" "test" -}} # I realise I only require one argument at the moment but I would still like to figure this out
{{ else if lookPath "python3" -}}
{{   $pythonCmd = "python3" -}}
{{ else if lookPath "python" -}}
{{   $pythonCmd = "python" -}}
{{ end -}}
{{ if ne "" $pythonCmd -}}
{{   $pythonArgs = mustCompact $pythonArgs -}}
[interpreters.py]
command = {{ $pythonCmd | quote }}
args = ...

{{ end -}}
```

But now I'm unsure abow how to best output the list of arguments as a TOML array of quoted strings which are comma+space separated.

Here are some of my approaches up to now with their results:

```
I: args = {{ $pythonArgs | toStrings }}
O: args = [-3 test test]

I: args = {{ $pythonArgs | toStrings | quote }}
O: args = "[-3 test test]"

I: args = {{ $pythonArgs | quote | toStrings }}
O: args = ["[-3 test test]"]

# Joining with `", "`, then quoting
I: args = [{{ $pythonArgs | join `", "` | quote }}]
O: args = ["-3\", \"test\", \"test"]

# Joining with `", "`, then manually wrapping with quotes
I: args = ["{{ $pythonArgs | join `", "` }}"]
O: args = ["-3", "test", "test"]

# Manually iterating, quoting and joining
# This isn't actually that complicated, but it looks horrific
I: args = [{{ range $i, $e := $pythonArgs }}{{ if gt $i 0 }}, {{ end }}{{ $e | quote }}{{ end }}]
O: args = ["-3", "test", "test"]
```

The previous 2 approaches when the argument list is empty (`$pythonArgs = list ""`):

```
I: args = ["{{ $pythonArgs | join `", "` }}"]
O: args = [""]

I: args = [{{ range $i, $e := $pythonArgs }}{{ if gt $i 0 }}, {{ end }}{{ $e | quote }}{{ end }}]
O: args = []
```

Both appear to be valid TOML, though I'm unsure if one is preferred over the other.

Is there a better way of achieving this that I might have missed?

I guess I need something equivalent to `| quoteEach | join ", "`.

**Comment by twpayne** [comment]
Sadly `text/template` is a pretty basic language (by intention of the designers), so an explicit `range` is probably needed.

An absolutely horrible idea is, and I feel dirty for even suggesting it, but if you know you have a slice of strings then you can use `toJson`:

```console
$ chezmoi execute-template '{{ list "-3" "test" "test" | toJson }}'
["-3","test","test"]
```
...there aren't spaces after the commas though :(

> I guess I need something equivalent to `| quoteEach | join ", "`.

Given that this is likely to be a common case (generating a list of quoted strings), I'd accept a PR that adds a `quoteEach` template function if you want to give it a go.

**Comment by bradenhilton** [comment, marked as answer]
As of [`v2.18.0`](https://github.com/twpayne/chezmoi/releases/tag/v2.18.0), this is now possible using the [`quoteList`](https://www.chezmoi.io/reference/templates/functions/quoteList/) template function (note the `[section]` example).

Thanks for implementing `quoteList` @twpayne.
